features/nvstore/source/nvstore_int_flash_wrapper.cpp

```diff
@@ -0,0 +1,192 @@
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define MIN(a,b)            ((a) < (b) ? (a) : (b))
+#define MAX(a,b)            ((a) > (b) ? (a) : (b))
+
+#define MAX_PAGE_SIZE   16
+
+static FlashIAP flash;
+
+static size_t get_page_size(void)
+{
+    return MIN(flash.get_page_size(), MAX_PAGE_SIZE);
+}
+
+// Program to Flash with alignments to page size
+// Parameters :
+// @param[in]   buffer - pointer to the buffer to be written
+// @param[in]   size - the size of the buffer in bytes.
+// @param[in]   address - the address of the internal flash, must be aligned to minimum writing unit (page size).
+// Return     : None.
+static int program_flash(size_t size, uint32_t address, const uint8_t *buffer)
+{
+    int ret;
+    uint32_t page_size, aligned_size, remainder;
+    uint8_t rem_buf[MAX_PAGE_SIZE];
+
+    page_size = get_page_size();
+    remainder = size % page_size;
+    aligned_size = size - remainder;
+    if (aligned_size) {
+        ret = flash.program(buffer, address, aligned_size);
+
+        if (ret) {
+            return -1;
+        }
+        address += aligned_size;
+        buffer += aligned_size;
+    }
+
+    if (!remainder) {
+        return 0;
+    }
+
+    memset(rem_buf, NVSTORE_BLANK_FLASH_VAL, page_size);
+    memcpy(rem_buf, buffer, remainder);
+    ret = flash.program(buffer, address, page_size);
+    if (ret) {
+        return -1;
+    }
+
+    return 0;
+}
+
+size_t nvstore_int_flash_get_sector_size(uint32_t address)
+{
+    return flash.get_sector_size(address);
+}
+
+int nvstore_int_flash_init(void)
+{
+    int ret;
+
+    ret = flash.init();
+    if (ret) {
+        return -1;
+    }
+
+    return 0;
+}
+
+int nvstore_int_flash_deinit(void)
+{
+    int ret;
+
+    ret = flash.deinit();
+    if (ret) {
+        return -1;
+    }
+
+    return 0;
+}
+
+int nvstore_int_flash_read(size_t size, uint32_t address, uint32_t *buffer)
+{
+    int ret;
+
+    if (!buffer) {
+        return -1;
+    }
+
+    if (!size)
+    {
+        return -1;
+    }
+    ret = flash.read(buffer, address, size);
+
+    if (ret) {
+        return -1;
+    }
+
+    return 0;
+}
+
+
+int nvstore_int_flash_erase(uint32_t address, size_t size)
+{
+    int ret;
+    size_t sector_size = nvstore_int_flash_get_sector_size(address);
+
+    if ((!size) || (size % sector_size) || (address % sector_size)) {
+        return -1;
+    }
+
+    // No need to iterate over sectors - Flash driver's erase API does it for us
+    ret = flash.erase(address, sector_size);
+    if (ret) {
+        return -1;
+    }
+
+    return 0;
+}
+
+
+int nvstore_int_flash_write(size_t size, uint32_t address, const uint32_t *buffer)
+{
+    int ret;
+    uint32_t page_size = get_page_size();
+    uint32_t sector_size = nvstore_int_flash_get_sector_size(address);
+    uint32_t chunk;
+    uint8_t *buf = (uint8_t *) buffer;
+
+    if ((!size) || (!buffer) || (address % page_size)) {
+        return -1;
+    }
+
+    while (size)
+    {
```

Comment:
on the line 172

Reply:
Please elaborate. Seems like a partial comment.

Reply:
`{` misplaced (as the rest of file has it fine)

Reply:
OK thanks.